# Project 03 -- Anthony Micco

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [2]:
# going to the Amazon directory and searching for all lines in tracks.csv containing "Little Mix"

In [3]:
%%bash
cd /anvil/projects/tdm/data/amazon/
pwd
ls

/anvil/projects/tdm/data/amazon
Reviews.csv
amazon_fine_food_reviews.csv
amazon_fine_food_reviews.sqlite
amazon_fine_food_reviews.zip
music.txt
spark
test
tracks.csv
tracks.db


In [4]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep "Little Mix" /anvil/projects/tdm/data/amazon/tracks.csv

TRAJZHI128F932BB39|"I'm A Little Mixed Up"|SONDRHG12AB017F90A|"South Side Lady (Live in Netherlands 1973) (Blues Reference)"|ARWF1971187B9B66CC|9ebffe22-c6e0-49bc-a6c2-f79a97bc38b3|"Koko Taylor"|222.04036|0.621432272125|0.423620810234|1969|6069564|-1|0
TROJQKV128F92D6730|"I'm A Little Mixed Up"|SOTNQQN12AAF3B2B46|"Rhythm & Blues Vol. 8"|ARJ0FXD1187B99F639|8b84d752-9f4e-498e-ad27-23a5de2b69c5|"Betty James"|165.27628|0.377304320601|0.203963938841|1993|3628911|-1|0


In [5]:
#finding who sang "Waiting for Blue" 

In [6]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep "Waiting for Blue" /anvil/projects/tdm/data/amazon/tracks.csv

TRQOXRL128F930D94A|"Waiting for Blue"|SOHFYOF12AB017FA1A|"Testing Superstition"|ARZGSBD1187B9919F2|2be0a553-3f67-4ee8-84d3-0824d9889006|"The June Spirit"|245.49832|0.481722049976|0.306129858047|2002|5459913|-1|0


In [7]:
#finding how many lines contain "Michael Jackson using the -c command

In [8]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -c "Michael Jackson" /anvil/projects/tdm/data/amazon/tracks.csv

253


In [9]:
#Counting how many lines contain "X Marks the Pedwalk"

In [10]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -c -i "X Marks the Pedwalk" /anvil/projects/tdm/data/amazon/tracks.csv

47


In [11]:
#Finding which line contains "Rainbow in the Night"

In [12]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -n -i "Rainbow in the Night" /anvil/projects/tdm/data/amazon/tracks.csv

351687:TRZHTHL128F42AB7ED|"Rainbow In The Night"|SOUOOBS12A8C13FA6C|"You Are Being Lied To"|ARQMX4G1187FB5934E|90f9fc57-7b6f-4229-a69f-84d2644882ff|Kinesis|214.80444|0.466848408257|0.360992335286|2005|3460954|-1|0


In [13]:
#Finding all the lines containing "garlic" 

In [14]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

TRHXDRN128F429F4BB|"Garlic Blues"|SONFGXK12A8C13FA75|"Sylvester Weaver Vol. 2 (1927)"|ARPJNT81187B98AD36|1e1fa4a5-3bd6-4a1c-9b1d-5c613a741500|"Sylvester Weaver"|180.47955|0.370703603519|0.283196596109|0|1871681|-1|0
TRCCKZJ128F92C4F7B|"Have Got Garlic Hanging On My Front Door"|SOYAHQF12A8C140C04|Signals|ARTUHAL1187B98B477|b93197b9-c613-4ecc-8146-4f7a11407f70|Marasma|400.79628|0.422275305174|0.215381650322|1998|3545788|-1|0
TRFQZIJ128F92FEEDB|"King Garlic"|SOWHCIV12AB0181A45|Spaceguerilla|AR5YT6E119B340313C|dccff302-f95e-46ce-8f95-a8ddd811b5d9|"Missus Beastly"|396.95628|0.389691318086|0.297792838565|0|4407167|-1|0
TRIYIKE128F423BB77|"Boeuf gros sel [Instrumental]"|SOHIEQA12A8AE46D0C|Garlick|ARUZB9Q1187FB45C9C|9ec13aa9-43e9-4a86-95be-d2b5c188dd20|"Hugues Aufray"|154.14812|0.561993937673|0.363323041772|1972|1401763|-1|0
TRTLHNN128F93273F6|"Garlic Bread"|SOSWOZF12AF72A3F30|"ASBO 4 Life"|ARJXVSV1187B9A252F|5167695e-3a2a-4311-9990-b7e2d13686a2|"Goldie Lookin Chain"|226.21995|0.640784077463|0

In [15]:
#Finding the lines that contain the whole word "Garlic"

In [16]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -w -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

TRHXDRN128F429F4BB|"Garlic Blues"|SONFGXK12A8C13FA75|"Sylvester Weaver Vol. 2 (1927)"|ARPJNT81187B98AD36|1e1fa4a5-3bd6-4a1c-9b1d-5c613a741500|"Sylvester Weaver"|180.47955|0.370703603519|0.283196596109|0|1871681|-1|0
TRCCKZJ128F92C4F7B|"Have Got Garlic Hanging On My Front Door"|SOYAHQF12A8C140C04|Signals|ARTUHAL1187B98B477|b93197b9-c613-4ecc-8146-4f7a11407f70|Marasma|400.79628|0.422275305174|0.215381650322|1998|3545788|-1|0
TRFQZIJ128F92FEEDB|"King Garlic"|SOWHCIV12AB0181A45|Spaceguerilla|AR5YT6E119B340313C|dccff302-f95e-46ce-8f95-a8ddd811b5d9|"Missus Beastly"|396.95628|0.389691318086|0.297792838565|0|4407167|-1|0
TRTLHNN128F93273F6|"Garlic Bread"|SOSWOZF12AF72A3F30|"ASBO 4 Life"|ARJXVSV1187B9A252F|5167695e-3a2a-4311-9990-b7e2d13686a2|"Goldie Lookin Chain"|226.21995|0.640784077463|0.462737056844|2009|4609633|-1|0
TRSAEGK128E0788EA2|Garlic|SOSIPTC12A6701C549|"Disappear Here"|ARX13A01187B99FF60|c76e4d1c-06ff-4594-b31c-4a93d9c1cb22|"Silver Sun"|188.47302|0.514594286426|0.362416410941|0|108

In [17]:
#Determing the number of lines that contain garlic as a substring and not a whole word

In [18]:
#First find the total number of times the word "garlic" appears in the file 

In [19]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -c -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

9


In [20]:
#Now find the total number of times the fullword (excluding substrings) "garlic" appears in the file

In [21]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -c -w -i "garlic" /anvil/projects/tdm/data/amazon/tracks.csv

6


In [22]:
#Subtract the two to find the total number of substrings 
print(9-6)

3


1.3) The June Spirit sings "Waiting for Blue" \
1.4) 253 lines contain "Michael Jackson" \
1.5) 47 lines contain "X Marks the Pedwalk" \
1.6) Line 351687 contains the "Rainbow in the Night \
1.9) There are 3 lines where "garlic" is a substring and not the world word

## Question 2

In [23]:
# Going to the Amazon directory and searching for all lines in the .csv files that contain "apple"

In [24]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -c -i "apple" /anvil/projects/tdm/data/amazon/*.csv

/anvil/projects/tdm/data/amazon/Reviews.csv:10635
/anvil/projects/tdm/data/amazon/amazon_fine_food_reviews.csv:10635
/anvil/projects/tdm/data/amazon/tracks.csv:1028


In [25]:
#Running the wildcard command on the movies_and_tv directory and figuring out what -OPTION command to fix it with

In [26]:
%%bash
cd /anvil/projects/tdm/data/movies_and_tv/
grep "toy story" /anvil/projects/tdm/data/movies_and_tv/*

grep: /anvil/projects/tdm/data/movies_and_tv/imdb2023: Is a directory
grep: /anvil/projects/tdm/data/movies_and_tv/imdb2024: Is a directory
grep: /anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes: Is a directory


/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes_reviews.csv:/m/1001063-anna_karenina,2008-04-05,Urban Cinefile,Fresh,,,"Garbo, who also starred in the 1927 silent version of the Tolstoy story, is radiant and vulnerable throughout the film, the centre of the emotional storm that engulfs her."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes_reviews.csv:/m/2_jacks,2013-10-17,Film Journal International,Rotten,,,"Granted, the Tolstoy story had the benefit of internal monologues that clarified motivations and impulses. But a good film finds ways of doing this through gesture, expression and other artful ways. This is not a good film."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes_reviews.csv:/m/annabelle_comes_home,2019-06-28,Popdust,Fresh,,,The scariest toy story of the year.
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes_reviews.csv:/m/bratz_the_movie,2007-09-15,Empire Magazine,Rotten,,,Lessons about friendship and tolerance in this live-action toy story are 

CalledProcessError: Command 'b'cd /anvil/projects/tdm/data/movies_and_tv/\ngrep "toy story" /anvil/projects/tdm/data/movies_and_tv/*\n'' returned non-zero exit status 2.

In [27]:
#Running the fixed wildcard command by adding -d recurse to the command

In [28]:
%%bash
cd /anvil/projects/tdm/data/movies_and_tv/
grep -d recurse -i "pointer" /anvil/projects/tdm/data/movies_and_tv/*

grep: /anvil/projects/tdm/data/movies_and_tv/imdb.db: binary file matches


/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0006451	1	Bobby Bumps and His Pointer Pup	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0006451	2	Bobby Bumps and His Pointer Pup	US	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0017829	11	The West Pointer	US	\N	working	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	10	El pointer	ES	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	13	The Pointer	\N	\N	original	\N	1
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	2	Como Treinar Um Pointer	BR	\N	\N	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	4	The Pointer	US	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0031803	8	El pointer	AR	\N	imdbDisplay	\N	0
/anvil/projects/tdm/data/movies_and_tv/imdb2023/title.akas.tsv:tt0146003	1	Pointe

grep: /anvil/projects/tdm/data/movies_and_tv/imdb2024.db: binary file matches


/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1112954-rookie,Chuck Wilson,False,L.A. Weekly,Rotten,,2002-06-01,"This is a boy's movie, a fact director John Lee Hancock ... and screenwriter Mike Rich ... make clear with such images as an anonymous boy and his dad in the stands watching the big game together, the father whispering pointers to his son."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1155467-glory_road,Matt Pais,False,Metromix.com,Fresh,3/4,2006-05-04,"A darn good story, told with enough conviction and impact to connect with people who don't know the difference between a three-pointer and a field goal."
/anvil/projects/tdm/data/movies_and_tv/rotten_tomatoes/rotten_tomatoes_critic_reviews.csv:m/1201402-17_again,Carrie Rickey,True,Philadelphia Inquirer,Fresh,3/4,2009-04-16,"For me, 17 Again is mildly diverting. For the tweens who are its target audience, Efron's outside shot is a three

2.2) There are **10635 lines** in /anvil/projects/tdm/data/amazon/Reviews.csv, **10635 lines** in /anvil/projects/tdm/data/amazon/amazon_fine_food_reviews.csv and **1028 lines** in /anvil/projects/tdm/data/amazon/tracks.csv that contain "apple". \
2.3) To make the command work you can add the **-d recurse** command which process directories using the action, recurse in our case, to read through the files in each directory as well as your current directory.

## Question 3

In [29]:
#redoing question 1.4 with the wc command to find the number of lines with "Michael Jackson"

In [30]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep "Michael Jackson" /anvil/projects/tdm/data/amazon/tracks.csv | wc

    253    2335   56549


In [31]:
#redoing question 1.5 with the wc command to find the number of lines with "X Marks the Pedwalk"

In [32]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -i "X Marks the Pedwalk" /anvil/projects/tdm/data/amazon/tracks.csv | wc

     47     291    7989


In [33]:
#Printing the head of the Reviews.csv file and only display ProfileName, Score and Summary

In [34]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head /anvil/projects/tdm/data/amazon/Reviews.csv

Id,ProductId,UserId,ProfileName,HelpfulnessNumerator,HelpfulnessDenominator,Score,Time,Summary,Text
1,B001E4KFG0,A3SGXH7AUHU8GW,delmartian,1,1,5,1303862400,Good Quality Dog Food,I have bought several of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells better. My Labrador is finicky and she appreciates this product better than  most.
2,B00813GRG4,A1D87F6ZCVE5NK,dll pa,0,0,1,1346976000,Not as Advertised,"Product arrived labeled as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was an error or if the vendor intended to represent the product as ""Jumbo""."
3,B000LQOCH0,ABXLMWJIXXAIN,"Natalia Corres ""Natalia Corres""",1,1,4,1219017600,"""Delight"" says it all","This is a confection that has been around a few centuries.  It is a light, pillowy citrus gelatin with nuts - in this case Filberts. And it is cut into tiny squares and then liberally co

In [35]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d',' -f 4,7,9

ProfileName,Score,Summary
delmartian,5,Good Quality Dog Food
dll pa,1,Not as Advertised
"Natalia Corres ""Natalia Corres""",4,"""Delight"" says it all"
Karl,2,Cough Medicine
"Michael D. Bigham ""M. Wassir""",5,Great taffy
Twoapennything,4,Nice Taffy
David C. Sullivan,5,Great!  Just as good as the expensive brands!
Pamela G. Williams,5,"Wonderful
R. James,5,Yay Barley


In [36]:
#Replacing all white spaces with underscores using the tr command in the music.txt file

In [37]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head /anvil/projects/tdm/data/amazon/music.txt | tr ' ' '_'

"I_love_this_CD.__So_inspiring!"
"Love_it!!__Great_seller!"
"I_bought_this_on_cassette_tape_in_the_80's._So_inspirational_to_me_back_then._Came_across_it_again_and_needed_to_be_uplifted!_Keith_Green's_music_still_works_magic._I_bought_CD's_for_my_girls_to_play_in_their_cars"
"as_good_as_i_remember_back_when_i_bought_the_one_in_the_80's"
"One_of_my_very_favourite_albums_from_one_of_my_very_favourite_singers.__I_was_happy_to_see_I_could_replace_the_old_worn_cassettes_from_years_ago."
"THank_you_Jesus_Lord_God,_that_brother_Green's_music_is_still_sounding_though_he_is_Home_with_you_now."
"I_recall_loving_his_other_albums_and_maybe_this_one_too,_forgot_but_I_figured_on_some_of_these_artists_seems_like_one_good_album_and_all_good_albums..especially_in_Christian_music..seemed_when_they_got_into_it_they_stayed_into_it_and_so_good_to_double_check_though_if_want_too_but_it_is_a_possible_very_good_album_because_I_usually_recall_if_too_bad_of_one_and_I_don't_on_this_one"
"Keith_Green's_music_are_

In [38]:
#Searching for lines containing chocolate from the first 76 lines of Reviews.csv

In [39]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head /anvil/projects/tdm/data/amazon/Reviews.csv

Id,ProductId,UserId,ProfileName,HelpfulnessNumerator,HelpfulnessDenominator,Score,Time,Summary,Text
1,B001E4KFG0,A3SGXH7AUHU8GW,delmartian,1,1,5,1303862400,Good Quality Dog Food,I have bought several of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells better. My Labrador is finicky and she appreciates this product better than  most.
2,B00813GRG4,A1D87F6ZCVE5NK,dll pa,0,0,1,1346976000,Not as Advertised,"Product arrived labeled as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was an error or if the vendor intended to represent the product as ""Jumbo""."
3,B000LQOCH0,ABXLMWJIXXAIN,"Natalia Corres ""Natalia Corres""",1,1,4,1219017600,"""Delight"" says it all","This is a confection that has been around a few centuries.  It is a light, pillowy citrus gelatin with nuts - in this case Filberts. And it is cut into tiny squares and then liberally co

In [40]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head -76 /anvil/projects/tdm/data/amazon/Reviews.csv | grep "chocolate" -i | cut -d, -f2

B00144C10S
B004N5KULM
B004N5KULM
B004N5KULM
B000E7VI7S
B0059WXJKM


3.1) There are 252 lines that contain "Michael Jackson" \
3.2) There are 47 lines that contain "X Marks the Pedwalk"

## Question 4

In [41]:
# printing the unique product ids that contain the words chocolate in the first 76 lines of Reviews.csv

In [42]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head -76 /anvil/projects/tdm/data/amazon/Reviews.csv | grep "chocolate" -i | cut -d, -f2 | uniq

B00144C10S
B004N5KULM
B000E7VI7S
B0059WXJKM


In [43]:
#determing the number of times each unique product id occurs

In [44]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head -76 /anvil/projects/tdm/data/amazon/Reviews.csv | grep "chocolate" -i | cut -d, -f2 | uniq -c

      1 B00144C10S
      3 B004N5KULM
      1 B000E7VI7S
      1 B0059WXJKM


In [45]:
#sorting the product ids from the previous step in descending order

In [46]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head -76 /anvil/projects/tdm/data/amazon/Reviews.csv | grep "chocolate" -i | cut -d, -f2 | uniq -c | sort -r  

      3 B004N5KULM
      1 B0059WXJKM
      1 B00144C10S
      1 B000E7VI7S


In [47]:
#Finding the 10 ProductIds most frequently associated with "chocolate"

In [48]:
%%bash
cd /anvil/projects/tdm/data/amazon/
cat /anvil/projects/tdm/data/amazon/Reviews.csv | grep "chocolate" -i | cut -d, -f2 | uniq -c | sort -r | head -n 10   

    259 B002NHYQAS
    228 B004U49QU2
    201 B001OCKIP0
    195 B001EQ55RW
    190 B004U43ZO0
    181 B0076MLL12
    181 B005K4Q68Q
    181 B005K4Q4KG
    181 B005K4Q1VI
    181 B005K4Q1T0


## Question 5

In [49]:
# finding the number of lines that contain the words love or hate

In [50]:
%%bash
cd /anvil/projects/tdm/data/amazon/
grep -d recurse -E -i "love|hate" /anvil/projects/tdm/data/amazon/Reviews.csv | wc -l

186028


In [51]:
#Printing all the lines that begin with a capital letter

In [52]:
%%bash
cd /anvil/projects/tdm/data/amazon/
cut -d',' -f2- /anvil/projects/tdm/data/amazon/Reviews.csv | grep -E '^[A-Z]' | head

ProductId,UserId,ProfileName,HelpfulnessNumerator,HelpfulnessDenominator,Score,Time,Summary,Text
B001E4KFG0,A3SGXH7AUHU8GW,delmartian,1,1,5,1303862400,Good Quality Dog Food,I have bought several of the Vitality canned dog food products and have found them all to be of good quality. The product looks more like a stew than a processed meat and it smells better. My Labrador is finicky and she appreciates this product better than  most.
B00813GRG4,A1D87F6ZCVE5NK,dll pa,0,0,1,1346976000,Not as Advertised,"Product arrived labeled as Jumbo Salted Peanuts...the peanuts were actually small sized unsalted. Not sure if this was an error or if the vendor intended to represent the product as ""Jumbo""."
B000LQOCH0,ABXLMWJIXXAIN,"Natalia Corres ""Natalia Corres""",1,1,4,1219017600,"""Delight"" says it all","This is a confection that has been around a few centuries.  It is a light, pillowy citrus gelatin with nuts - in this case Filberts. And it is cut into tiny squares and then liberally coated with

In [53]:
#printing all the lines in the head that start with a capital letter in the Summary column

In [55]:
%%bash
cd /anvil/projects/tdm/data/amazon/
head -n 20 /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d',' -f9 | grep '^[A-Z]'

Summary
Good Quality Dog Food
Not as Advertised
Cough Medicine
Great taffy
Nice Taffy
Great!  Just as good as the expensive brands!
Yay Barley
Healthy Dog Food
The Best Hot Sauce in the World
My Cats Are Not Fans of the New Food
Strawberry Twizzlers - Yummy
Love it!
GREAT SWEET CANDY!


In [56]:
#counting all the lines that start with a capital letter in the Summary column

In [58]:
%%bash
cd /anvil/projects/tdm/data/amazon/
cat /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d',' -f9 | grep '^[A-Z]' | wc -l

439383


In [59]:
#Counting the lines that end with the word "great" in the Summary column

In [62]:
%%bash
cd /anvil/projects/tdm/data/amazon/
cat /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d',' -f9 | grep 'great$' | wc -l

2683


In [63]:
#Finding how many reviews contain two digits 

In [65]:
%%bash
cd /anvil/projects/tdm/data/amazon/
cat /anvil/projects/tdm/data/amazon/Reviews.csv | cut -d',' -f10 | grep '[0-9].*[0-9]' | wc -l

60831


5.2)There are **186028 lines** that contain the words "love" or "hate" in the amazon Reviews.csv directory \
5.5)There are **439383 lines** that begin with a capital letter in the Summary column
5.6)There are **2683 lines** that end with "great" in the Summary column
5.7)There are **60831 lines** that contain two digits in the Text column

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.